src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/AbiphoneContactHelper.java

```diff
@@ -0,0 +1,204 @@
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.event.entity.player.ItemTooltipEvent;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import org.jetbrains.annotations.Nullable;
+import org.lwjgl.input.Keyboard;
+
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class AbiphoneContactHelper {
+
+	private static final AbiphoneContactHelper INSTANCE = new AbiphoneContactHelper();
+
+	public static AbiphoneContactHelper getInstance() {
+		return INSTANCE;
+	}
+
+	private final Pattern patternUsualLocation = Pattern.compile("(§5§o)?§7Usual location: §b(.*)");
+	private String selectedWaypointName = "";
+	private long lastClick = 0L;
+
+	@SubscribeEvent
+	public void onItemTooltip(ItemTooltipEvent event) {
+		if (!NotEnoughUpdates.INSTANCE.hasSkyblockScoreboard()) return;
+		if (!Utils.getOpenChestName().equals("Contacts Directory")) return;
+
+		List<String> list = event.toolTip;
+		if (list == null) return;
+		if (list.isEmpty()) return;
+		String rawNpcName = list.get(0);
+		String npcName = StringUtils.cleanColour(rawNpcName);
+
+		JsonObject data = getJsonData(npcName);
+		if (data == null) return;
+
+		if (NotEnoughUpdates.INSTANCE.config.tooltipTweaks.abiphoneContactRequirements) {
+			if (data.has("requirement")) {
+				JsonArray requirements = data.get("requirement").getAsJsonArray();
+				if (requirements.size() > 0) {
+					list.add(" ");
+					list.add("§e§lRequirements:");
+
+					for (JsonElement requirementObject : requirements) {
+						String requirement = requirementObject.getAsString();
+						list.add(requirement);
+					}
+				}
+			}
+		}
+
+		if (NotEnoughUpdates.INSTANCE.config.tooltipTweaks.abiphoneContactMarker) {
+			if (data.has("x")) {
+				if (selectedWaypointName.equals(npcName)) {
+					list.set(0, npcName + " §f- §aMarker set!");
+				}
+
+				list.add(" ");
+				if (selectedWaypointName.equals(npcName)) {
+					list.add("§eClick to remove the marker!");
+				} else {
+					list.add("§eClick to set a marker!");
+				}
+				list.add("§eShift-Click to teleport to the nearest waypoint!");
+			}
+		}
+	}
+
+	@SubscribeEvent
+	public void onStackClick(SlotClickEvent event) {
+		if (!NotEnoughUpdates.INSTANCE.hasSkyblockScoreboard()) return;
+		if (!NotEnoughUpdates.INSTANCE.config.tooltipTweaks.abiphoneContactMarker) return;
+		if (!Utils.getOpenChestName().equals("Contacts Directory")) return;
+
+		ItemStack stack = event.slot.getStack();
+		if (stack == null || stack.getDisplayName() == null) return;
+
+		String rawNpcName = stack.getDisplayName();
+		String npcName = StringUtils.cleanColour(rawNpcName);
+
+		JsonObject data = getJsonData(npcName);
+		if (data == null) return;
+		if (!data.has("x")) return;
+
+		int x = data.get("x").getAsInt();
+		int y = data.get("y").getAsInt();
+		int z = data.get("z").getAsInt();
+
+		String islandName = fetchIslandName(stack);
```

Comment:
even if this works i would probably prefer having the island in the repo

Reply:
effort

Reply:
It works and as jani noted, it's more effort for him. Or is your argument about performance? We can cache the whole lore

Reply:
> It works and as jani noted, it's more effort for him. Or is your argument about performance? We can cache the whole lore

I added the island in my draft repo pr, if you want to you can change it in code